File: junior/contest2/prog/string_stats.java
import java.io.*;
import java.util.*;

public class string_stats {
	public static void main(String[] args) throws IOException {
		Scanner input = new Scanner(System.in);
		String s = input.nextLine();
		StringBuilder sb = new StringBuilder(s); // the mutable copy
		int[] diffs = new int[26]; // case insensitive counting array
		int vowels = 0;
		int up = 0;
		for(int i = 0; i < s.length(); ++i) {
			char c = s.charAt(i);
			if(Character.isUpperCase(c)) {
				++diffs[c-'A'];
				switch(c) {
				case 'A':
				case 'E':
				case 'I':
				case 'O':
				case 'U':
					++vowels;
					break;
				}
				++up;
			} else if(Character.isLowerCase(c)) {
				++diffs[c-'a'];
				switch(c) {
				case 'a':
				case 'e':
				case 'i':
				case 'o':
				case 'u':
					++vowels;
					break;
				}
			} else { // non-letter
				sb.setCharAt(i, ' '); // replace with a space
			}
		}
		int cnt = 0; int max_val = 0;
		for(int i = 0; i < 26; ++i) {
			if(diffs[i] != 0)
				++cnt;
			max_val = Math.max(max_val, diffs[i]);
		}
		System.out.println(cnt);
		System.out.println(vowels);
		System.out.println(up);
		System.out.println(max_val);

		ArrayList<String> words = new ArrayList<String>();
		Scanner is = new Scanner(sb.toString());
		do {
			String word = is.next();
			if(word.isEmpty())
				break;
			if(words.isEmpty()) {
				words.add(word);
				continue;
			}
			if(word.length() < words.get(words.size()-1).length())
				continue; // skip shorter word
			if(word.length() == words.get(words.size()-1).length()) {
				words.add(word);
				continue; // add same length word
			}
			words.clear(); // empty storage for new longer length
			words.add(word);
		} while(is.hasNext());

		if(words.size() == 1) {
			System.out.println(words.get(0));
			return;
		}

		ArrayList<String> orgs = new ArrayList<String>(words);
		for(int i = 0; i < words.size(); ++i)
			words.get(i).toLowerCase(); // Convert to lower case for sorting
		Collections.sort(words);
		String first = words.get(0);
		for(int i = 0; i < words.size(); ++i)
			if(words.get(i).compareTo(first) == 0) {
				System.out.println(orgs.get(i)); // Output the one before conversion
				return;
			}
	}//end of main()
}//end of class

File: junior/contest2/prog/string_stats.cpp
#include <bits/stdc++.h>
using namespace std;

void tolower(string& s) {
	for(auto& c : s)
		if(isupper(c))
			c = tolower(c);
}

int main() {
	string s;
	getline(cin, s);
	int diffs[26] = {};
	int vowels = 0;
	int up = 0;
	for(char& c : s) { // Use Counting Array and case insensitive
		if(isupper(c)) {
			++diffs[c-'A'];
			switch(c) {
			case 'A':
			case 'E':
			case 'I':
			case 'O':
			case 'U':
				++vowels;
				break;
			}
			++up;
		} else if(islower(c)) {
			++diffs[c-'a'];
			switch(c) {
			case 'a':
			case 'e':
			case 'i':
			case 'o':
			case 'u':
				++vowels;
				break;
			}
		}
		else { // non-letter
			c = ' '; // replace with a space
		}
	}
	int cnt = 0; int max_val = 0;
	for(int i = 0; i < 26; ++i) {
		if(diffs[i])
			++cnt;
		max_val = max(max_val, diffs[i]);
	}
	cout << cnt << endl;
	cout << vowels << endl;
	cout << up << endl;
	cout << max_val << endl;

	vector<string> words;
	istringstream is(s);
	do {
		string word;
		is >> word; //getline(is, word, ' ');
		if(word.empty())
			break;
		if(words.empty()) {
			words.push_back(word);
			continue;
		}
		if(word.length() < words.back().length())
			continue; // skip shorter word
		if(word.length() == words.back().length()) {
			words.push_back(word);
			continue; // add same length word
		}
		words.clear(); // empty storage for new longer length
		words.push_back(word);
	} while(is);

	if(words.size() == 1) {
		cout << words[0] << endl;
		return 0;
	}

	auto low_words = words;
	for(auto& word : low_words)
		tolower(word); // Convert to lower case for sorting
	auto sorted = low_words;
	sort(sorted.begin(), sorted.end());
	auto first = sorted[0];
	for(int i = 0; i < low_words.size(); ++i)
		if(low_words[i] == first) {
			cout << words[i] << endl; // Output the one before conversion
			return 0;
		}
	return 0;
}
